### DISCUSSION TITLE
Why should I use Pinia instead of Classes?

### DISCUSSION THREAD

**Discussion by varna** [vuejs/pinia#831] [Help and Questions] [answered]
Why should anyone use Pinia instead of JS or TS Classes?

**Comment by posva** [comment, marked as answer]
https://pinia.esm.dev/introduction.html#why-should-i-use-pinia

**Comment by varna** [comment]
> https://pinia.esm.dev/introduction.html#why-should-i-use-pinia

OK. It adds DX and SSR. Awesome, I love it! But what I actually meant to ask is:

> Why should Pinia use `defineStore` pattern instead of `Class`

We already used `vuex` for large scale applications and know that `plugins` and `mixins` is the wrong way to code.
Having `classes`, `interfaces` and `generics` would allow us to write maintainable code.

Example of store in my work project which annoys the hell out of me:
```js
import items from '../getters/_items'
import actions from '../getters/_actions'
import stock from '../getters/_stock'
import pending from '../getters/_pending'
import cost from '../getters/_cost'
import price from '../getters/_price'

const type = `chemical`

export default {
  namespaced: true,
  getters: {
    items: items(type),
    purchases: actions(type, 'purchases'),
    sell: actions(type, 'sell'),
    write_off: actions(type, 'write_off'),
    stock: stock(type),
    pending: pending(type),
    cost: cost(type),
    price: price(type),
    ...
}
```

It just feels like reinventing the wheel. Wouldn't this make more sense? I stole and updated the Cart from example: https://stackblitz.com/edit/github-eyireq?file=src/models/cart.js

Maybe we could create a DX, SSR friendly class called Pinia?
```js
class Cart extends Pinia {
  // and now our cart has DX and SSR support
}
```

**Reply by posva** [reply]
Maybe, yeah. It could also be a plugin. The state would likely need to go in a function though (instead of just being class properties), being quite unnatural and therefore missing the main point of looking like classes. This is necessary because the state properties must be created inside of an effectScope for everything to work.

**Reply by yukulele** [reply]
> OK. It adds DX and SSR.

What is DX?

**Reply by varna** [reply]
Well... 2 years later I'm glad that I was wrong. **Pinia is the greatest store library**.

1. Simple API
2. Very _very_ flexible. I'm very glad about `useStore` pattern. 

@yukulele  DX = Developer Experience.

Thank You @posva !

**Reply by yukulele** [reply]
I'm sorry to reopen the debate, but I'm not sure I understand what Pinia offers beyond the use of classes.

Why this:
```ts
// src/stores/counter.ts
import { ref, computed } from "vue";
import { defineStore } from "pinia";

export const useCounterStore = defineStore("counter", () => {
  const count = ref(0);
  const doubleCount = computed(() => count.value * 2);

  function increment() {
    count.value++;
  }

  return { count, doubleCount, increment };
});

```
```html
<!-- src/components/TheCounter.vue-->
<script setup lang="ts">
import { useCounterStore } from '@/stores/counter.js';
const counterStore = useCounterStore()
</script>

<template>
  <div>count: {{ counterStore.count }}</div>
  <div>double: {{ counterStore.doubleCount }}</div>
  <button @click="counterStore.increment">+</button>
</template>
```
is better than this:
```ts
// src/stores/counter.ts
import { ref, computed } from "vue";

class CounterStore {
  count = ref(0);
  doubleCount = computed(() => this.count.value * 2);

  increment() {
    this.count.value++;
  }
}

export const counterStore = new CounterStore();
```
```html
<!-- src/components/TheCounter.vue-->
<script setup lang="ts">
import { counterStore } from '@/stores/counter.js';
</script>

<template>
  <div>count: {{ counterStore.count }}</div>
  <div>double: {{ counterStore.doubleCount }}</div>
  <button @click="counterStore.increment">+</button>
</template>
```
?

**Reply by ilajosmanov** [reply]
I ask myself the same question :)
I mostly wanna use classes as services for my components (like in angular)

**Reply by pavi2410** [reply]
I agree with @yukulele. I love Pinia, but class is more familiar to me as it's a standard language feature.

@yukulele's class example can be improved like this:

```ts
// src/stores/counter.ts
import { ref, computed } from "vue";

class CounterStore {
  count = ref(0);

  // use native class getters
  get doubleCount() {
    return this.count.value * 2;
  }

  increment() {
    this.count.value++;
  }
}

// counterStore is now a Pinia store.
export const counterStore = pinia.createFromClass(new CounterStore());
```

**Reply by SkinnyArman** [reply]
@varna Sooo what was your experience that you found out Pinia is the best way for global state management?
I have previously used Vuex for a large scale application and the result was meh. If I were to go back I would have used just plain JS to create a singleton service like in Angular and pass it down to components using provide/inject.

**Reply by ghost** [reply]
> Well... 2 years later I'm glad that I was wrong. **Pinia is the greatest store library**.
> 
> 1. Simple API
> 2. Very _very_ flexible. I'm very glad about `useStore` pattern.
> 
> @yukulele DX = Developer Experience.
> 
> Thank You @posva !

it's horrible, classes are nice, and composables in vue are bad a step back. It parse a lot of code to find that composables and eat a huge amount of pc resources, but in classes all data is already  ready to use 


That are nice but not perfect https://github.com/PentaTea/pinia-class-component
